Mathlib/Tactic/Linter/CommandStart.lean

```diff
@@ -0,0 +1,326 @@
+  `Mathlib.Meta.setBuilder,
+
+  -- # misc exceptions
+
+  -- We ignore literal strings.
+  `str,
+
+  -- list notation, the pretty-printer prefers `a :: b`
+  ``«term_::_»,
+
+  -- negation, the pretty-printer prefers `¬a`
+  ``«term¬_»,
+
+  -- declaration name, avoids dealing with guillemets pairs `«»`
+  ``Parser.Command.declId,
+
+  `Mathlib.Tactic.superscriptTerm, `Mathlib.Tactic.subscript,
+
+  -- notation for `Bundle.TotalSpace.proj`, the total space of a bundle
+  -- the pretty-printer prefers `π FE` over `π F E` (which we want)
+  `Bundle.termπ__,
+
+  -- notation for `Finset.slice`, the pretty-printer prefers `𝒜 #r` over `𝒜 # r` (mathlib style)
+  `Finset.«term_#_»,
+
+  -- The docString linter already takes care of formatting doc-strings.
+  ``Parser.Command.docComment,
+
+  -- `omit [A] [B]` prints as `omit [A][B]`, see https://github.com/leanprover/lean4/pull/8169
+  ``Parser.Command.omit,
+  ]
+
+/--
+Given an array `a` of `SyntaxNodeKind`s, we accumulate the ranges of the syntax nodes of the
+input syntax whose kind is in `a`.
+
+The linter uses this information to avoid emitting a warning for nodes with kind contained in
+`unlintedNodes`.
+-/
+def getUnlintedRanges(a : Array SyntaxNodeKind) :
+    Std.HashSet String.Range → Syntax → Std.HashSet String.Range
+  | curr, s@(.node _ kind args) =>
+    let new := args.foldl (init := curr) (·.union <| getUnlintedRanges a curr ·)
+    if a.contains kind then
+      new.insert (s.getRange?.getD default)
+    else
+      new
+  | curr, _ => curr
+
+/-- Given a `HashSet` of `String.Ranges` `rgs` and a further `String.Range` `rg`,
+`outside rgs rg` returns `true` if and only if `rgs` contains a range the completely contains
+`rg`.
+
+The linter uses this to figure out which nodes should be ignored.
+-/
+def outside? (rgs : Std.HashSet String.Range) (rg : String.Range) : Bool :=
+  let superRanges := rgs.filter fun {start := a, stop := b} => (a ≤ rg.start && rg.stop ≤ b)
+  superRanges.isEmpty
+
+/-- `mkWindow orig start ctx length` extracts from `orig` a string that starts at the first
+non-whitespace character before `start`, then expands to cover `length + ctx` more characters
+and continues still until the first non-whitespace character.
+
+In essence, it extracts the substring of `orig` that begins at `start`, continues for `length + ctx`
+plus expands left and right until it encounters the first whitespace character, to avoid
+cutting into "words".
+
+*Note*. `start` is the number of characters *from the right* where our focus is!
+-/
+def mkWindow (orig : String) (start ctx length : Nat) : String :=
+  let head := orig.dropRight (start + 1) -- `orig`, up to one character before the discrepancy
+  let middle := orig.takeRight (start + 1) --
+  let headCtx := head.takeRightWhile (!·.isWhitespace)
+  let tail := middle.drop (length + ctx) |>.takeWhile (!·.isWhitespace)
+  s!"{headCtx}{middle.take (length + ctx)}{tail}"
```

Comment:
It's getting late and my head is knotting up, trying to understand this: can you add some unit tests for this function to document what it does?

Reply:
I added some tests.